openmicroscopy/openmicroscopy
# Short url (rebase metadata 5.1)

Original work started in https://github.com/openmicroscopy/openmicroscopy/pull/4401 but due to database changes separate PR is required for 5.1

-----

This PR is a proposal for https://trello.com/c/sJCSdD7P/174-redirect-service

To test:

- create short url for ``/webclient/?show=image-1`` (currently stored as omero.model.Alias)

 ```
c=omero.client('localhost', 4064)
sess=c.createSession('root', 'secret')
conn=omero.gateway.BlitzGateway(client_obj=c)
update = conn.getUpdateService()
ns = "host/port/prefix" #or referrer
short = omero.model.AliasI()
short.setNs(omero.rtypes.rstring(ns))
short.origin = omero.rtypes.rstring("/webclient/?show=image-1")
short.details.permissions = omero.model.PermissionsI("rwr-r-")
short = update.saveAndReturnObject(short)
print short.id.val, short.origin.val
print 'short', base62.from_decimal(short.id.val)
 ```

- go to https://host:port/prefix/url/_SHORT_
- check if you were redirected to the expected location

Example: Alias:ID = 1
```
$ curl -I http://localhost:4080/url/B
HTTP/1.0 302 FOUND
Date: Wed, 20 Jan 2016 13:40:14 GMT
Server: WSGIServer/0.1 Python/2.7.10
Vary: Cookie
Content-Type: text/html; charset=utf-8
Location: http://localhost:4080/webclient/?show=image-1
```
------

DB upgrade patch:

```
$ createdb -E UTF8 -O omero omero_metadata_5.1
$ psql -h localhost -U omero omero_metadata_5.1 < OMERO5.1__2.sql 
Password for user omero: 
BEGIN
...
UPDATE 1
COMMIT
$ omero_start
No descriptor given. Using etc/grid/default.xml
Waiting on startup. Use CTRL-C to exit
$ ./build.py -f components/tools/OmeroWeb/build.xml test -DTEST=test/integration/test_redirect.py
...
============================= test session starts ==============================
platform darwin -- Python 2.7.10 -- py-1.4.31 -- pytest-2.7.0 -- /usr/local/opt/python/bin/python2.7
rootdir: /Users/ola/OMERO/ome.git/components/tools, inifile: pytest.ini
collecting ... collected 1 items
test/integration/test_redirect.py::TestShortUrl::test_short_url PASSED
 generated xml file: /Users/ola/OMERO/ome.git/components/tools/OmeroWeb/target/reports/junit-results.xml 
=========================== 1 passed in 3.22 seconds ===========================
```


```
$ createdb -E UTF8 -O omero omero_metadata_5.1
$ psql -h localhost -U omero omero_metadata_5.1 < OMERO5.1__1.sql 
Password for user omero: 
BEGIN
...
UPDATE 1
COMMIT
$ psql -h localhost -U omero omero_metadata_5.1 < sql/psql/OMERO5.1__2/OMERO5.1__1.sql 
Password for user omero: 
BEGIN
CREATE FUNCTION
 omero_assert_db_version 
-------------------------
 
(1 row)

DROP FUNCTION
INSERT 0 1
CREATE TABLE
CREATE TABLE
CREATE TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE SEQUENCE
INSERT 0 1
CREATE SEQUENCE
INSERT 0 1
CREATE FUNCTION
CREATE TRIGGER
CREATE TRIGGER
CREATE TRIGGER
UPDATE 1
                               status                                
---------------------------------------------------------------------
                                                                    +
                                                                    +
                                                                    +
 YOU HAVE SUCCESSFULLY UPGRADED YOUR DATABASE TO VERSION OMERO5.1__2+
                                                                    +
                                                                    +
 
(1 row)

COMMIT
$ omero_start
No descriptor given. Using etc/grid/default.xml
Waiting on startup. Use CTRL-C to exit
$ ./build.py -f components/tools/OmeroWeb/build.xml test -DTEST=test/integration/test_redirect.py
...
============================= test session starts ==============================
platform darwin -- Python 2.7.10 -- py-1.4.31 -- pytest-2.7.0 -- /usr/local/opt/python/bin/python2.7
rootdir: /Users/ola/OMERO/ome.git/components/tools, inifile: pytest.ini
collecting ... collected 1 items
test/integration/test_redirect.py::TestShortUrl::test_short_url PASSED
 generated xml file: /Users/ola/OMERO/ome.git/components/tools/OmeroWeb/target/reports/junit-results.xml 
=========================== 1 passed in 4.73 seconds ===========================
```
